### PR TITLE
Raise error when torch.load a storage on a non-existing device

### LINKED ISSUE #7178: Should torch.load and torch.save take a device (since PyTorch 0.4)
Reading the [PyTorch 0.4.0 migration guide](https://pytorch.org/2018/04/22/0_4_0-migration-guide.html) I came across the [`torch.device`](https://pytorch.org/docs/stable/tensor_attributes.html) abstraction.

It is great for creating a device once and then passing it around and using the `.to` functions, e.g. as in:

```python
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
inputs = data.to(device)
```

The [`torch.load`](https://github.com/pytorch/pytorch/blob/d564ecb4a515e34184c976617f57b2eb01665660/torch/serialization.py#L241) function does not take a device, though. Instead it takes a `map_location` argument which can either be a lambda, a mapping, or since https://github.com/pytorch/pytorch/pull/4203 it can be a string like `'cpu'`.

Now the question is why are there these two different concepts and can they be unified into one device abstraction? Otherwise we can pass the device around _except_ for serialization where we need to transform the device abstraction into a `map_location` parameter.

Can we unify these concepts behind an API like the following?

```python
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
restored = torch.load('model.pth', device=device)
```

Related: https://github.com/pytorch/pytorch/issues/6630 - `torch.save` should also take a device

### PR DESCRIPTION
**Pull request by zou3519**
Before, doing torch.load(...) on a CUDA tensor on a CPU-only machine
would raise an unreadable error:

```
~/pytorch/pytorch/torch/cuda/__init__.py in __enter__(self)
    223         if self.idx is -1:
    224             return
--> 225         self.prev_idx = torch._C._cuda_getDevice()
    226         if self.prev_idx != self.idx:
    227             torch._C._cuda_setDevice(self.idx)

AttributeError: module 'torch._C' has no attribute '_cuda_getDevice'
```

This PR makes it so that torch.load raises a hard error if one tries to
load a storage onto a non-existing device and suggests the user to use
torch.load's map_location feature.

Fixes #7178, #6630

### CODE REVIEW

File: torch/serialization.py [hunk 1]
```diff
@@ -70,6 +70,19 @@ def _cuda_deserialize(obj, location):
             device = 0
         else:
             device = max(int(location[5:]), 0)
+
+        if not torch.cuda.is_available():
+            raise RuntimeError('Attempting to deserialize object on a CUDA '
+                               'device but torch.cuda.is_available() is False. '
+                               'If you are running on a CPU-only machine, '
+                               'please use torch.load with map_location to map '
```
Comment (apaszke) [inline torch/serialization.py]: Can you suggest `map_location='cpu'` here?

File: test/test_cuda.py [hunk 2]
```diff
@@ -1099,6 +1100,14 @@ def test_cat_bad_input_sizes(self):
         z = torch.randn(2, 2, 1).cuda()
         self.assertRaises(RuntimeError, lambda: torch.cat([x, y, z], dim=1))
 
+    @unittest.skipIf(torch.cuda.device_count() >= 10, "Loading a cuda:9 tensor")
+    @unittest.skipIf(not PY3, "Tensor was serialized with Python 3")
+    def test_load_nonexistent_device(self):
+        msg = r'Attempting to deserialize object on CUDA device 9'
+        with self.assertRaisesRegex(RuntimeError, msg):
+            fname = os.path.join(os.path.dirname(__file__), 'data/pi-cuda9.pt')
```
Comment (apaszke) [inline test/test_cuda.py]: Hmm is there any way to fake saving this without checking in those binary files in the repo? `git` is pretty bad at handling those.

Comment (zou3519) [reply test/test_cuda.py]: Possibly, I could write an output file on CPU and then replace all instances of 'cpu' with 'cuda:0' and/or 'cuda:9'. Let me try that and see how it goes

**Comment by zou3519** [comment]
@apaszke I was able to fake a tensor for one of the tests (the cuda:9 one). It's harder to fake a cuda tensor given the serialized output of a cpu tensor, so I just copied the bytes into the test file to avoid having a binary file. Let me know what you think

**Comment by li-roy** [comment]
I think your current approach is better than overwriting `torch.cuda.is_available`.

**Review by soumith — APPROVED** [review]

**Pull request merged**